ipc: add dynamically allocated buffers to icmsg

doki-nordic opened the pull request:
The `icmsg` and `icmsg_me` backends have limitations in context of concurrent access to a single instance. Some of the limitations are:

* sending by two or more thread at the same time may cause -EBUSY
* allocating TX buffer will cause errors in other threads that want to allocate before first thread sent the message,
* during no-copy receive, when RX buffer is on hold receiving is totally blocked.

This backend resolves those limitations by adding dynamically allocated buffers on shared memory next to ICmsg circular buffer. The data is passed using those buffers, so concurrency is not a problem. The ICmsg is used only to pass short 2-byte messages containing references to those buffers. The backend also supports multiple endpoint. The ipc/icmsg_me sample was modified to support this backend.

dchat-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+}
+
+/**
+ * Release received buffer. This function will just sends release message over ICMsg.
```
```suggestion
 * Release received buffer. This function will just send release message over ICMsg.
```

dchat-nordic reviewed (commented):
Looks logically good, just grammar.

carlocaione reviewed (dismissed):
Please, separate the core changes from the sample ones.

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+	sys_cache_data_flush_range(block, size + BLOCK_HEADER_SIZE);
+
+	r = icmsg_send_wrapper(dev_data, remote_addr, tx_block_index);
+
```
Unnecessary blank line

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+ * Find endpoint that was registered with name that matches name
+ * contained in the endpoint bound message received from remote. This function
+ * must be called when mutex is locked.
+ *
```
`dev_data` ?

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+ * Find registered endpoint that matches given "bound endpoint" message. When found,
+ * remote endpoint address is saved for future use and "release bound endpoint" message
+ * is send. This function must be called when mutex is locked.
+ *
```
`dev_data` ?

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+ *
+ * @retval 0	match not found
+ * @retval 1	match found and processing was successful
+ * @retval < 0	error occurred during processing
```
`retval < 0` ?

doki-nordic replied:
Ok, it is not valid doxygen syntax. I will fix that.

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+ *
+ * @param[in] ept	Endpoint to use
+ *
+ * @return		O or negative error code
```
```suggestion
 * @retval		O or negative error code
```

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+
+/**
+ * Send bound message on specified endpoint.
+ *
```
`dev_data`?

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+	alloc_size = msg_len;
+	r = alloc_tx_buffer(dev_data, &alloc_size, &buffer, K_FOREVER);
+	if (r >= 0) {
+		msg = (struct ept_bound_msg *)buffer;
```
Shouldn't `struct ept_bound_msg` be packed or does it msg 4 byte aligned?

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+	size_t i;
+	int r = 0;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
```
Maybe configurable timeout could be an option?

doki-nordic replied:
I disagree. It is only to avoid access to the same memory by more threads. It is not long running procedure, so in such case, I don't see point of using something different then K_FOREVER.

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+		goto exit;
+	}
+
+	addr_or_msg_type = message[0];
```
Could we use defines or static const variables to describe indexes here?

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+	return 0;
+}
+
+/**
```
I don't sure if you should use doxygen mark inside source file

KAGA164 reviewed (commented):
Overall looks promising. I putted a few comments.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+			/* Wait for releasing if there is no enough space and exit loop
+			 * on timeout
+			 */
+			r = k_sem_take(&dev_data->block_wait_sem, timeout);
```
The `k_timeout_t` holds absolute time. It is calculated in macro that converts e.g. ms to k_timeout_t. You can read some more there: https://docs.zephyrproject.org/latest/kernel/services/timing/clocks.html#subsystems-using-k-timeout-t

I was using the second example for this implementation.

KAGA164 replied:
Hmm I not sure that it is absolute without doing additional operation

doki-nordic commented on samples/subsys/ipc/ipc_service/multi_endpoint/README.rst:
```diff
@@ -0,0 +1,76 @@
+
+    $ west build -b nrf5340dk_nrf5340_cpuapp -- \
+      -DDTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpuapp_icmsg_w_buf.overlay \
+      -DDTC_OVERLAY_FILE_REMOTE=boards/nrf5340dk_nrf5340_cpunet_icmsg_w_buf.overlay
```
It is not done autoamtically by the build system, so I added handling of this varable in Cmake: https://github.com/zephyrproject-rtos/zephyr/pull/58741/files#diff-1810cb21049c8e65908e16a8b62cb8c03e98484bbdc00c3511c856e8b3f783acR28

I was following naming of the variable that was already there `BOARD_REMOTE`

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+ * @param[out] size		Size of the allocated buffer if not NULL.
+ *				The size is also checked if it fits in the blocks area.
+ *				If it is NULL, no size validation is performed.
+ * @return		Block index or negative error code
```
`@retval` is for specific value,  `@return` is for general description of the function result, so I am thinking it is right in this context.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+
+/**
+ * Allocate buffer for transmission
+ *
```
I intentionally skipped `dev_data` and `conf` parameters documentation in entire file. It is always the same in this file, and it is always obvious what it is, and the comments are for developer that creates/modified the file (not a user of the file). I can add it, but I don't see the point.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+ * @param[in,out] size	Required size of the buffer. If zero, first available block is
+ *			allocated and all subsequent available blocks. Size actually
+ *			allocated which is not less than requested.
+ * @param[out] block	First allocated block. Block header contains actually allocated
```
Right, it is `buffer`. I forget to update the comment.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1222 @@
+	uint8_t message[2] = { addr_or_msg_type, tx_block_index };
+	int r;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
```
I don't see the point. Putting 2 bytes into ICMsg queue is not long or blocking operation, so we will not block here to often. Even if we block, we will serve higher priority thread fast, because of priority Inheritance on mutex. I don't see a real use case when something different than K_FOREVER makes sense.

emob-nordic commented:
I think the documentation similar to `./doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst` should be added, explaining how this works.

hubertmis reviewed (commented):
Unfortunately, I won't review this fully anytime soon due to other more urgent tasks.

A few potential problems came to my mind while looking through this code:
1. Cache management - I see it is implemented
2. Keeping memory allocation metadata in the same memory space as the allocated data itself may be considered a security issue. But as long as this is just for point-to-point communication, it's probably fine. Though some security experts could have a look
3. Yet another hundreds-lines dynamic memory allocation module in Zephyr could be a huge maintenance cost. I'm curious if one of the existing allocation modules like memory slabs could be used here instead.

eobalski reviewed (commented):
Why is new backend implemented instead of using multi endpoint variant? What's the rationale?

Edit:
Can't we use heap implementation from zephyr instead of implementing new allocator? Not sure about the requirements for the backend but i'm happy to hear.

doki-nordic commented:
@hubertmis, @emob-nordic, answering to your questions:

> Why is new backend implemented instead of using multi endpoint variant? What's the rationale?

The rationale is at the beginning of the PR description. The multiendpoint backend has **significant limitations** that cannot be resolved without completely rewriting it. The purpose of this PR is to provide backend that has no significant limitations and can be used in almost in any use case (maybe except some restricted low footprint or very high performance cases).

> 2. Keeping memory allocation metadata in the same memory space as the allocated data itself may be considered a security issue. But as long as this is just for point-to-point communication, it's probably fine. Though some security experts could have a look

No, memory allocation metadata is not placed in shared memory. One exception is the buffer length (which must be shared). I payed a lot of attention to always verify it before use to make it secure. I was implementing it as a secure (non-hackable) communication channel. I agree, security experts could have a look. Can you suggest someone?

> 3. Yet another hundreds-lines dynamic memory allocation module in Zephyr could be a huge maintenance cost. I'm curious if one of the existing allocation modules like memory slabs could be used here instead.

> Can't we use heap implementation from zephyr instead of implementing new allocator? Not sure about the requirements for the backend but i'm happy to hear.

It is not a new hundreds-lines dynamic memory allocation module. I am actually using one of the zephyr's kind of allocators: `sys_bitarray_alloc`/`sys_bitarray_free`. Most of the code is endpoint bounding, data passing, e.t.c. The allocation and deallocation is contained in two functions that has total of 140 lines of code and a lot of those lines are thread synchronizing/signaling, security and error checks which will not go away if I switch to a different zephyr's allocator.

I cannot use slabs (or others) because of security reasons (see my answer above). Only alternative to zephyr's `sys_bitarray_alloc` is `sys_mem_blocks_alloc_contiguous`. I was considering it, but `sys_mem_blocks_alloc_contiguous` actually calls `sys_bitarray_alloc` and convert block index to pointer. In this case, my module must take the pointer and calculate it back to first block index and allocated block count, which is pointless and my allocation code will actually grow. As you can see, `sys_bitarray_alloc` is only zephy'r allocator that can be used here.

hubertmis commented:
Fair enough. Thanks for responding to my concerns!

peknis commented on samples/subsys/ipc/ipc_service/multi_endpoint/README.rst:
```diff
@@ -0,0 +1,77 @@
+Changing the backend
+********************
+
+You can change the backend to ``icmsg_with_buf`` by changing the devicetree
+overlay files.
```
```suggestion
To change the backend to ``icmsg_with_buf``, switch the devicetree
overlay files as follows:
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+################################################
+
+This backend is built on top of the :ref:`ipc_service_backend_icmsg`.
+Data transferred over this backend travels in dynamically allocated
+buffers on shared memory. The ICMsg is used to send just references to
+those buffers. It also supports multiple endpoints.
```
```suggestion
Data transferred over this backend travels in dynamically allocated buffers on shared memory.
The ICMsg just sends references to the buffers.
It also supports multiple endpoints.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+Overview
+========
+
+The shared memory is divided into two parts. One is reserved for the ICMsg and
+the other contains equal-sized blocks. The number of blocks is configured in
+the devicetree.
```
```suggestion
The shared memory is divided into two parts.
One is reserved for the ICMsg and the other contains equal-sized blocks.
The number of blocks is configured in the devicetree.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+
+The data sending process is following:
+
+* The sender allocates one or more blocks. If there are not enough sequential
+  blocks, it waits using the timeout provided in the parameter which also
+  includes K_FOREVER and K_NO_WAIT.
```
```suggestion
* The sender allocates one or several blocks.
  If there are not enough sequential blocks, it waits using the timeout provided in a parameter (including K_FOREVER and K_NO_WAIT).
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+* The sender allocates one or more blocks. If there are not enough sequential
+  blocks, it waits using the timeout provided in the parameter which also
+  includes K_FOREVER and K_NO_WAIT.
+* The allocated blocks are filled with data. For the zero-copy case, this is
+  done by the caller, otherwise, it is copied automatically.
+  During this time other threads are not blocked in any way as long as there are
+  enough free blocks for them. They can allocate, send data and receive data.
```
```suggestion
* The allocated blocks are filled with data.
  For the zero-copy case, this is done by the caller, otherwise, it is copied automatically.
  During this time, other threads are not blocked in any way as long as there are
  enough free blocks for them.
  They can allocate, send, and receive data.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+  done by the caller, otherwise, it is copied automatically.
+  During this time other threads are not blocked in any way as long as there are
+  enough free blocks for them. They can allocate, send data and receive data.
+* A message containing block index is sent over ICMsg to the receiver.
+  The size of the ICMsg queue is big enough to hold messages for all blocks, so
+  it will never overflow.
```
```suggestion
* A message containing the block index is sent over ICMsg to the receiver.
  The size of the ICMsg queue is large enough to hold messages for all blocks, so
  it will never overflow.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+* A message containing block index is sent over ICMsg to the receiver.
+  The size of the ICMsg queue is big enough to hold messages for all blocks, so
+  it will never overflow.
+* The receiver can hold the data as long as he wants. Again, other threads are
+  not blocked as long as there are enough free blocks for them.
```
```suggestion
* The receiver can hold the data as long as desired.
  Again, other threads are not blocked as long as there are enough free blocks for them.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+  it will never overflow.
+* The receiver can hold the data as long as he wants. Again, other threads are
+  not blocked as long as there are enough free blocks for them.
+* When data is no longer needed, backed sends a release message over ICMsg.
```
```suggestion
* When data is no longer needed, the backend sends a release message over ICMsg.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+* The receiver can hold the data as long as he wants. Again, other threads are
+  not blocked as long as there are enough free blocks for them.
+* When data is no longer needed, backed sends a release message over ICMsg.
+* When backed receives this message, it deallocates blocks. It is done
+  internally by the backend and it is invisible to the caller.
```
```suggestion
* When the backend receives this message, it deallocates all blocks.
  It is done internally by the backend and it is invisible to the caller.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+Configuration
+=============
+
+The backend is configured via Kconfig and devicetree.
```
```suggestion
The backend is configured using Kconfig and devicetree.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+The backend is configured via Kconfig and devicetree.
+When configuring the backend, do the following:
+
+* Define two memory regions and assign them to ``tx-region`` and ``rx-region``
+  of an instance. Ensure that the memory regions used for data exchange are
+  unique (not overlapping any other region) and accessible by both domains
+  (or CPUs).
```
```suggestion
* Define two memory regions and assign them to ``tx-region`` and ``rx-region`` of an instance.
  Ensure that the memory regions used for data exchange are unique (not overlapping any other region) and accessible by both domains (or CPUs).
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+  (or CPUs).
+* Define the number of allocable blocks for each region with ``tx-blocks`` and
+  ``rx-blocks``.
+* Define MBOX devices which are used to send the signal that informs the other
+  domain (or CPU) that data has been written. Ensure that the other domain
+  (or CPU) is able to receive the signal.
```
```suggestion
* Define MBOX devices for sending a signal that informs the other domain (or CPU) of the written data.
  Ensure that the other domain (or CPU) can receive the signal.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+   };
+
+
+You must provide a similar configuration for the other side of the
+communication (domain or CPU) but you must swap the MBOX channels, memory
+regions (``tx-region`` and ``rx-region``) and block count (``tx-blocks`` and
+``rx-blocks``).
```
```suggestion
You must provide a similar configuration for the other side of the communication (domain or CPU).
Swap the MBOX channels, memory regions (``tx-region`` and ``rx-region``) and block count (``tx-blocks`` and
``rx-blocks``).
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg_with_buf.rst:
```diff
@@ -0,0 +1,93 @@
+buffers on shared memory. The ICMsg is used to send just references to
+those buffers. It also supports multiple endpoints.
+
+This architecture allows for overcoming some common problems
+with other backends (mostly related to multithread access and zero-copy).
+That is why this backend provides an alternative that has no known significant
+limitations like the others.
```
```suggestion
This architecture allows for overcoming some common problems with other backends (mostly related to multithread access and zero-copy).
This backend provides an alternative with no significant limitations.
```

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+
+	/* Set flag that ICMsg is bounded and now, endpoint bounding may start. */
+	atomic_or(&dev_data->flags, FLAG_ICMSG_BOUNDED);
+	schedule_ept_bound_process(dev_data);
```
Does it make sense to trigger bound process on no initiator side on ICMsg bound since it will be triggered on every bound message reception?

doki-nordic replied:
But the user can call bound of endpoint before ICMsg is bounded (or later), that is always scheduled when both are conditions are true.

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+	*token = ept;
+
+	/* Rest of the bounding will be done in the system workqueue. */
+	schedule_ept_bound_process(dev_data);
```
Is it needed on no initiator side?

doki-nordic replied:
Yes

KAGA164 replied:
Could we just assume that endpoint registration is done after icmsg control channel is open. It could simplify logic because we wouldn't need to have states and bound message could be sent directly from this function.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+
+	if (size != NULL) {
+		if (invalidate_cache) {
+			sys_cache_data_invd_range(block, BLOCK_HEADER_SIZE);
```
Assuming that:
1) the receiver does not write to received buffer (it is ensured on language level by using `const`)
2) the receiver does not flush the cache with received buffer (he don't need to, because he did not write anything)
then there is no need to align to the cache line.

That's a big if, so I implemented a cache aligned blocks that are configurable and enabled by default.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+			   BLOCK_HEADER_SIZE));
+
+	return icmsg_open(&conf->icmsg_config, &dev_data->icmsg_data, &cb,
+			  (void *)instance);
```
The casing is not because of `void`. It is because the parameter takes simple pointer and the `instance` is pointer to const.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+
+	/* Reserve new endpoint index. */
+	ept_index = atomic_inc(&dev_data->flags) & FLAG_EPT_COUNT_MASK;
+	if (ept_index >= NUM_EPT) {
```
1. I cannot move before atomic_inc, because it will no longer be "atomic".
2. IMO, assertion is the best solution here. The endpoint are not something dynamic. There is specific number of endpoints in the application, so if you reach the limit then it is not a runtime error. This is a wrong configuration and for this case assert is the best.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+	struct backend_data *dev_data = instance->data;
+
+	dev_data->conf = conf;
+	dev_data->is_initiator = (conf->rx.blocks_ptr < conf->tx.blocks_ptr);
```
Why? Doing it automatically saves a lot of trouble for user, because he don't need to make sure that he put right values for each core. Things gets even more complicated when you have more than one core. What if you want to talk with one core as initiator and with the other not?

Detecting it in runtime makes the bounding a little bit bigger, because program contains support for both cases. I check it when I was implementing it. It was few hundreds bytes (I don't remember exact number now).

In my plans for improvements I have initiator detection in compile time, so during the preprocessing we will know if we should compile code for initiator, follower or both. I didn't implemented it to not over-complicate first version.

KAGA164 replied:
I was just wonder if this condition will work in feature when maybe different processors could have different memory mapping but I don't sure if such case is even possible. My other question is if we even need to have roles in this module. It seems to me that both sides could send bound message with local endpoint address to each others.

doki-nordic commented:
Review finding fixed.

KAGA164 commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+	case MSG_RELEASE_DATA:
+		ept_addr = EPT_ADDR_INVALID;
+		__fallthrough;
+	case MSG_RELEASE_BOUND:
```
Is there any difference between `MSG_RELEASE_BOUND` and `MSG_RELEASE_DATA`? Could we simplify it just to `MSG_RELEASE`

doki-nordic replied:
https://github.com/zephyrproject-rtos/zephyr/pull/58741#discussion_r1357951890

eobalski reviewed (changes requested):
To increase code readability i suggest to explicitly call functions responsible for control messages with `control_` prefix or anything that would help understand what is imcsg used for.
This goes for `received` and`bound` callbacks and more, like `icmsg_send_wrapper` -> `send_control_message` (maybe).

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+struct backend_data {
+	const struct icmsg_with_buf_config *conf; /* Backend instance config. */
+	struct icmsg_data_t icmsg_data;	/* ICMsg data. */
+	struct k_mutex mutex;		/* Mutex to protect: ICMsg send call and
```
Actually the reasons for this mutex are in your message:

> however the implementation does not allow to use `K_FOREVER`.

That is why I added the mutex.

doki-nordic commented on subsys/ipc/ipc_service/backends/Kconfig.icmsg_with_buf:
```diff
@@ -0,0 +1,41 @@
+	bool "ICMSG backend with dynamically allocated buffers"
+	default y
+	depends on MBOX
+	depends on DT_HAS_ZEPHYR_IPC_ICMSG_WITH_BUF_ENABLED
```
I disagree. Those dependencies are not needed, since the I am using define from the SPSC that is always available regardless of those configs.

I don't think that DT binding is better. If I do it that way, the user have to take care of the alignment. Why adding more unnecessary work for the user, more responsibilities, since it can be automatically done by inheriting value from SPSC. This is always the same memory, so alignment is the same.

doki-nordic replied:
I will remove IPC_SERVICE_BACKEND_ICMSG_WITH_BUF_CACHE_ALIGN (always enabled). This module will always inherit SPSC.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+ * the backend calls "bound" callback and sends back "Release bound endpoint".
+ * The follower saves the endpoint address. The follower's endpoint is ready to send
+ * and receive data. When the initiator gets "Release bound endpoint" message or any
+ * data messages, it calls bound endpoint and it is ready to send data.
```
When bound message was processed, the receiver sends MSG_RELEASE_BOUND, so the sender knows that bounding was finished and can call the `bound` callback.

doki-nordic replied:
I will refactor the receiving functions to make more clear what's the difference between those type of packets.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+};
+
+struct icmsg_message {
+	uint8_t msg_type;	/* Message type. */
+	uint8_t ept_addr;	/* Endpoint address or zero for MSG_RELEASE_DATA. */
+	uint8_t block_index;	/* Block index to send or release. */
```
I moved block count to the DT instead of kconfig, because I assume different instances talks with different CPUs. In "kconfig" case, all CPUs would need the same block count (or size). I think that we need more flexibility, since there is no almost additional cost putting it in DT.

Regarding `block count` vs `block size`, I don't have very strong arguments for any of the those approaches. They both have advantages and disadvantages. I simply choose the approach that was more natural for me.

I think that runtime block size exchange is not a good solution. I see no benefits over DT option except user provides block count just on one side. The problem is that a lot of code will be added to handle that. Currently most of block count/size/address calculations are done in build time.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+	}
+
+	/* Clear bit. If cleared, specific block will not be hold after the callback. */
+	sys_bitarray_clear_bit(conf->rx_hold_bitmap, rx_block_index);
```
If sender does not correctly allocates the buffers, then it is a malfunction and entire link is corrupted, so there is no point of checking it. Only reason for checks in "malfunction" situation is security. In this case we don't have any security risk.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+					 */
+	struct k_work ep_bound_work;	/* Work item for bounding processing. */
+	struct k_sem block_wait_sem;	/* Semaphore for waiting for free blocks. */
+	struct ept_data ept[NUM_EPT];	/* Array of registered endpoints. */
```
You thinking about allocating it on the heap?

I followed the same pattern from other backends that number of endpoints is kconfig option. I don't know why this backed should be different. Also, list require going over all previous items before I reach the one that I need. I prefer to use array.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+#define WAITING_BOUND_MSG_EMPTY 0xFFFF
+
+/** Number of bytes per each ICMsg message. It is used to calculate size of ICMsg area. */
+#define BYTES_PER_ICMSG_MESSAGE 8
```
Actually, emob-nordic explained in https://github.com/zephyrproject-rtos/zephyr/pull/58741#discussion_r1258038402

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+
+struct icmsg_with_buf_config {
+	struct icmsg_config_t icmsg_config;	/* Configuration of the ICMsg. */
+	struct channel_config rx;		/* RX channel config. */
```
No, entire `icmsg_with_buf_config` is const and filled during compile time, so there is no point of adding more levels of pointers there.

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+			sys_cache_data_invd_range(block, BLOCK_HEADER_SIZE);
+			__sync_synchronize();
+		}
+		allocable_size = ch_conf->block_count * ch_conf->block_size;
```
I see no benefits of that. It is not used in more places than here.

doki-nordic commented:
KAGA164 wrote:

> I was just wonder if this condition will work in feature when maybe different processors could have different memory mapping but I don't sure if such case is even possible.

I will not work if the shared memory will be mapped in some strange way. We can reconsider this approach.

> My other question is if we even need to have roles in this module. It seems to me that both sides could send bound message with local endpoint address to each others.

If you go back in history of this PR, at the beginning the protocol was symmetric. The bounding procedure was more complicated because we have no guarantees which side will start bounding first, what will be delay between bounding related packets, the implementation needed more states and more code to handle that. There was a race condition in old implementation that I had to fix. During fixing I noticed that asymmetric protocol requires much less code, so I switched to asymmetric.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1251 @@
+#define BLOCK_ALIGNMENT sizeof(size_t)
+
+/** Number of bytes per each ICMsg message. It is used to calculate size of ICMsg area. */
+#define BYTES_PER_ICMSG_MESSAGE 8
```
In the future `???` which represents packet header len in `spsc_puf` will be exported to header file. For now lets stick with `sizeof(uint32_t)`

doki-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c:
```diff
@@ -0,0 +1,1281 @@
+	MSG_DATA = 0,		/* Data message. */
+	MSG_RELEASE_DATA,	/* Release data buffer message. */
+	MSG_BOUND,		/* Endpoint bounding message. */
+	MSG_RELEASE_BOUND,	/* Release endpoint bound message. */
```
I can rename the `MSG_RELEASE_BOUND` to be more descriptive.

eobalski reviewed (commented):
I can suggest ic**b**msg since its blocked allocator based on icmsg.

doki-nordic commented:
I pushed new changes:
* all review findings fixed
* rebased to the latest commit
* miltiendpoint sample switched to sysbuild.

peknis commented on samples/subsys/ipc/ipc_service/multi_endpoint/README.rst:
```diff
@@ -0,0 +1,74 @@
+To change the backend to ``icbmsg``, switch the devicetree
+overlay files as follows:
+
+.. code-block:: console
+
+    $ west build -b nrf5340dk_nrf5340_cpuapp --sysbuild -- \
+      -DDTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpuapp_icbmsg.overlay \
+      -Dremote_DTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpunet_icbmsg.overlay
```
```suggestion
.. code-block:: console

   west build -b nrf5340dk_nrf5340_cpuapp --sysbuild -- \
   -DDTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpuapp_icbmsg.overlay \
   -Dremote_DTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpunet_icbmsg.overlay
```

peknis commented on samples/subsys/ipc/ipc_service/multi_endpoint/README.rst:
```diff
@@ -0,0 +1,74 @@
+Open a serial terminal (for example Minicom or PuTTY) and connect the board with the
+following settings:
+
+- Speed: 115200
+- Data: 8 bits
+- Parity: None
+- Stop bits: 1
```
```suggestion
* Speed: 115200
* Data: 8 bits
* Parity: None
* Stop bits: 1
```

peknis commented on samples/subsys/ipc/ipc_service/multi_endpoint/README.rst:
```diff
@@ -0,0 +1,74 @@
+   :board: nrf5340dk_nrf5340_cpuapp
+   :goals: debug
+
+Open a serial terminal (for example Minicom or PuTTY) and connect the board with the
+following settings:
```
```suggestion
Open a serial terminal (for example Minicom or PuTTY) and connect the board with the following settings:
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+
+
+You must provide a similar configuration for the other side of the communication (domain or CPU).
+Swap the MBOX channels, memory regions (``tx-region`` and ``rx-region``)
+and block count (``tx-blocks`` and ``rx-blocks``).
```
```suggestion
Swap the MBOX channels, memory regions (``tx-region`` and ``rx-region``) and block count (``tx-blocks`` and ``rx-blocks``).
```

peknis replied:
```suggestion
Swap the MBOX channels, memory regions (``tx-region`` and ``rx-region``), and block count (``tx-blocks`` and ``rx-blocks``).
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+The ICMsg just sends references to the buffers.
+It also supports multiple endpoints.
+
+This architecture allows for overcoming some common problems with
+other backends (mostly related to multithread access and zero-copy).
```
```suggestion
This architecture allows for overcoming some common problems with other backends (mostly related to multithread access and zero-copy).
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+The data sending process is following:
+
+* The sender allocates one or more blocks.
+  If there are not enough sequential blocks,
+  it waits using the timeout provided in the parameter which also includes K_FOREVER and K_NO_WAIT.
```
```suggestion
  If there are not enough sequential blocks, it waits using the timeout provided in the parameter that also includes K_FOREVER and K_NO_WAIT.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+* The sender allocates one or more blocks.
+  If there are not enough sequential blocks,
+  it waits using the timeout provided in the parameter which also includes K_FOREVER and K_NO_WAIT.
+* The allocated blocks are filled with data. For the zero-copy case,
+  this is done by the caller, otherwise, it is copied automatically.
```
```suggestion
* The allocated blocks are filled with data.
  For the zero-copy case, this is done by the caller, otherwise, it is copied automatically.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+  During this time other threads are not blocked in any way as long as there are enough free blocks for them.
+  They can allocate, send data and receive data.
+* A message containing the block index is sent over ICMsg to the receiver.
+  The size of the ICMsg queue is large enough to hold messages for all blocks,
+  so it will never overflow.
```
```suggestion
  The size of the ICMsg queue is large enough to hold messages for all blocks, so it will never overflow.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+  so it will never overflow.
+* The receiver can hold the data as long as desired.
+  Again, other threads are not blocked as long as there are enough free blocks for them.
+* When data is no longer needed, the backed sends a release message over ICMsg.
```
```suggestion
* When data is no longer needed, the backend sends a release message over ICMsg.
```

peknis commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -0,0 +1,88 @@
+When configuring the backend, do the following:
+
+* Define two memory regions and assign them to ``tx-region`` and ``rx-region`` of an instance.
+  Ensure that the memory regions used for data exchange are unique (not overlapping any other region)
+  and accessible by both domains (or CPUs).
```
```suggestion
  Ensure that the memory regions used for data exchange are unique (not overlapping any other region) and accessible by both domains (or CPUs).
```

eobalski reviewed (approved):
Just some small nits.

emob-nordic commented on samples/subsys/ipc/ipc_service/multi_endpoint/README.rst:
```diff
@@ -0,0 +1,74 @@
+
+    $ west build -b nrf5340dk_nrf5340_cpuapp --sysbuild -- \
+      -DDTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpuapp_icbmsg.overlay \
+      -Dremote_DTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpunet_icbmsg.overlay
```
Add this to `sample.yaml`
```
  sample.ipc.multi_endpoint.icbmsg:
    platform_allow: nrf5340dk_nrf5340_cpuapp
    integration platforms:
      - nrf5340dk_nrf5340_cpuapp
    tags: ipc
    sysbuild: true
    extra_args:
      DTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpuapp_icbmsg.overlay
      remote_DTC_OVERLAY_FILE=boards/nrf5340dk_nrf5340_cpunet_icbmsg.overlay

doki-nordic committed:
ipc: add dynamically allocated buffers to icmsg

The icmsg and icmsg_me backends has limitations in context of
concurrent access to single instance. Some limitations are:
* sending by more thread at the same time may cause -EBUSY
* allocating TX buffer will cause errors in other threads that want
  to allocate before first thread sent the message,
* during no-copy receive, when RX buffer is on hold receiving is
  totally blocked.
This backend resolves those limitations by adding dynamically allocated
buffers on shared memory next to ICmsg circular buffer. The data is
passed using those buffers, so concurrency is not a problem. The ICmsg
is used only to pass short 2-byte messages containing references to
those buffers. The backend also supports multiple endpoint.
The ipc/icmsg_me sample was modified to support this backend.

Signed-off-by: Dominik Kilian <Dominik.Kilian@nordicsemi.no>

doki-nordic committed:
samples: ipc: Modify icmsg_me sample allowing any multiendpoint backend

There are more ipc_service backends that supports multiple endpoint.
This sample can be used for any of those backends, so this commits
makes the sample more generic. The default backend it still icmsg_me,
but now, the sample has files and instructions for icmsg_with_buf
backend.

Signed-off-by: Dominik Kilian <Dominik.Kilian@nordicsemi.no>
